File: ww-ui/src/utils/logger.ts
/**
 * Log levels for the client-side logger
 * Higher levels include all lower level logs
 * ERROR (0) - Only critical errors
 * WARN (1) - Warnings and errors
 * INFO (2) - Important events (connections, chat, announcements) + warn/error
 * DEBUG (3) - Everything including high-frequency events (player input, game state)
 */
export enum LogLevel {
  ERROR = 0,
  WARN = 1,
  INFO = 2,
  DEBUG = 3,
}

const levelNames: Record<LogLevel, string> = {
  [LogLevel.ERROR]: "ERROR",
  [LogLevel.WARN]: "WARN",
  [LogLevel.INFO]: "INFO",
  [LogLevel.DEBUG]: "DEBUG",
};

const levelFromString: Record<string, LogLevel> = {
  error: LogLevel.ERROR,
  warn: LogLevel.WARN,
  info: LogLevel.INFO,
  debug: LogLevel.DEBUG,
};

class Logger {
  private level: LogLevel;

  constructor() {
    // Default to INFO, can be overridden by environment variable
    const envLevel = process.env.NEXT_PUBLIC_LOG_LEVEL?.toLowerCase();
    this.level =
      envLevel && levelFromString[envLevel] !== undefined
        ? levelFromString[envLevel]
        : LogLevel.INFO;
  }

  setLevel(level: LogLevel): void {
    this.level = level;
  }

  setLevelFromString(levelStr: string): boolean {
    const level = levelFromString[levelStr.toLowerCase()];
    if (level !== undefined) {
      this.level = level;
      return true;
    }
    return false;
  }

  getLevel(): LogLevel {
    return this.level;
  }

  getLevelName(): string {
    return levelNames[this.level];
  }

  private shouldLog(level: LogLevel): boolean {
    return level <= this.level;
  }

  private formatMessage(level: LogLevel, message: string): string {
    const timestamp = new Date().toISOString();
    return `[${timestamp}] [${levelNames[level]}] ${message}`;
  }

  /**
   * Debug logs for high-frequency, verbose messages
   * Use for: player input, game state updates, internal state changes
   */
  debug(message: string, ...args: unknown[]): void {
    if (this.shouldLog(LogLevel.DEBUG)) {
      console.log(this.formatMessage(LogLevel.DEBUG, message), ...args);
    }
  }

  /**
   * Info logs for important events
   * Use for: player connect/disconnect, chat messages, announcements
   */
  info(message: string, ...args: unknown[]): void {
    if (this.shouldLog(LogLevel.INFO)) {
      console.log(this.formatMessage(LogLevel.INFO, message), ...args);
    }
  }

  /**
   * Warn logs for warning-level messages
   * Use for: deprecated API usage, recoverable errors, unknown message types
   */
  warn(message: string, ...args: unknown[]): void {
    if (this.shouldLog(LogLevel.WARN)) {
      console.warn(this.formatMessage(LogLevel.WARN, message), ...args);
    }
  }

  /**
   * Error logs for critical failures
   * Use for: failures that need attention, unrecoverable errors
   */
  error(message: string, ...args: unknown[]): void {
    if (this.shouldLog(LogLevel.ERROR)) {
      console.error(this.formatMessage(LogLevel.ERROR, message), ...args);
    }
  }
}

// Export a singleton instance
export const logger = new Logger();

// Export for convenience
export default logger;

File: pkg/logger/logger.go
package logger

import (
	"fmt"
	"log"
	"os"
	"strings"
	"sync"
)

// Level represents log verbosity levels
// Higher levels include all lower level logs
// ERROR (0) - Only critical errors
// WARN (1) - Warnings and errors
// INFO (2) - Important events (connections, chat, announcements) + warn/error
// DEBUG (3) - Everything including high-frequency events (player_input, game_state)
type Level int

const (
	LevelError Level = iota
	LevelWarn
	LevelInfo
	LevelDebug
)

var levelNames = map[Level]string{
	LevelError: "ERROR",
	LevelWarn:  "WARN",
	LevelInfo:  "INFO",
	LevelDebug: "DEBUG",
}

var levelFromString = map[string]Level{
	"error": LevelError,
	"warn":  LevelWarn,
	"info":  LevelInfo,
	"debug": LevelDebug,
}

type Logger struct {
	mu     sync.RWMutex
	level  Level
	stdLog *log.Logger
	errLog *log.Logger
}

var defaultLogger = &Logger{
	level:  LevelInfo,
	stdLog: log.New(os.Stdout, "", log.LstdFlags),
	errLog: log.New(os.Stderr, "", log.LstdFlags),
}

// SetLevel sets the global log level
func SetLevel(level Level) {
	defaultLogger.mu.Lock()
	defer defaultLogger.mu.Unlock()
	defaultLogger.level = level
}

// SetLevelFromString sets log level from a string (error, warn, info, debug)
func SetLevelFromString(levelStr string) error {
	level, ok := levelFromString[strings.ToLower(levelStr)]
	if !ok {
		return fmt.Errorf("invalid log level: %s (valid: error, warn, info, debug)", levelStr)
	}
	SetLevel(level)
	return nil
}

// GetLevel returns the current log level
func GetLevel() Level {
	defaultLogger.mu.RLock()
	defer defaultLogger.mu.RUnlock()
	return defaultLogger.level
}

// GetLevelName returns the name of the current log level
func GetLevelName() string {
	return levelNames[GetLevel()]
}

// shouldLog returns true if the given level should be logged
func shouldLog(level Level) bool {
	defaultLogger.mu.RLock()
	defer defaultLogger.mu.RUnlock()
	return level <= defaultLogger.level
}

// Debug logs debug-level messages (high frequency, verbose)
// Use for: player_input, game_state updates, internal state changes
func Debug(format string, args ...any) {
	if shouldLog(LevelDebug) {
		defaultLogger.stdLog.Printf("[DEBUG] "+format, args...)
	}
}

// Info logs info-level messages (important events)
// Ex: player connect/disconnect, chat messages, announcements
func Info(format string, args ...any) {
	if shouldLog(LevelInfo) {
		defaultLogger.stdLog.Printf("[INFO] "+format, args...)
	}
}

// Warn logs warning-level messages
// Ex: deprecated API usage, recoverable errors, unknown message types
func Warn(format string, args ...any) {
	if shouldLog(LevelWarn) {
		defaultLogger.stdLog.Printf("[WARN] "+format, args...)
	}
}

// Error logs error-level messages
// Ex: failures that need attention, unrecoverable errors
func Error(format string, args ...any) {
	if shouldLog(LevelError) {
		defaultLogger.errLog.Printf("[ERROR] "+format, args...)
	}
}

// Fatal logs an error and exits
func Fatal(format string, args ...any) {
	defaultLogger.errLog.Printf("[FATAL] "+format, args...)
	os.Exit(1)
}

// ParseLevel converts a string to a Level
func ParseLevel(s string) (Level, error) {
	level, ok := levelFromString[strings.ToLower(s)]
	if !ok {
		return LevelInfo, fmt.Errorf("invalid log level: %s", s)
	}
	return level, nil
}
